lib/teleterm/vnet/service.go

```diff
@@ -14,35 +14,263 @@
+		// to the user.
+
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		if s.status == statusRunning {
+			s.status = statusNotRunning
+		}
+	}()
+
+	s.processManager = processManager
+	s.status = statusRunning
 	return &api.StartResponse{}, nil
 }
 
+// Stop stops VNet and cleans up used resources. Blocks until VNet stops or ctx is canceled.
 func (s *Service) Stop(ctx context.Context, req *api.StopRequest) (*api.StopResponse, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.stopLocked()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
 	return &api.StopResponse{}, nil
 }
 
-// Close stops the current VNet instance and prevents new instances from being started.
-//
+func (s *Service) stopLocked() error {
+	if s.status == statusClosed {
+		return trace.CompareFailed("VNet service has been closed")
+	}
+
+	if s.status == statusNotRunning {
+		return nil
+	}
+
+	s.processManager.Close()
+	err := s.processManager.Wait()
+	if err != nil && !errors.Is(err, context.Canceled) {
+		return trace.Wrap(err)
+	}
+
+	s.status = statusNotRunning
+	return nil
+}
+
+// Close stops VNet service and prevents it from being started again. Blocks until VNet stops.
 // Intended for cleanup code when tsh daemon gets terminated.
 func (s *Service) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := s.stopLocked()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.status = statusClosed
 	return nil
 }
+
+type appProvider struct {
+	daemonService      *daemon.Service
+	clientStore        *client.Store
+	insecureSkipVerify bool
+}
+
+func (p *appProvider) ListProfiles() ([]string, error) {
+	profiles, err := p.clientStore.ListProfiles()
+	return profiles, trace.Wrap(err)
+}
+
+func (p *appProvider) GetCachedClient(ctx context.Context, profileName, leafClusterName string) (*client.ClusterClient, error) {
+	uri := uri.NewClusterURI(profileName).AppendLeafCluster(leafClusterName)
+	client, err := p.daemonService.GetCachedClient(ctx, uri)
+	return client, trace.Wrap(err)
+}
+
+func (p *appProvider) ReissueAppCert(ctx context.Context, profileName, leafClusterName string, app types.Application) (tls.Certificate, error) {
+	clusterURI := uri.NewClusterURI(profileName).AppendLeafCluster(leafClusterName)
+	cluster, _, err := p.daemonService.ResolveClusterURI(clusterURI)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	client, err := p.daemonService.GetCachedClient(ctx, clusterURI)
+	if err != nil {
+		return tls.Certificate{}, trace.Wrap(err)
+	}
+
+	// TODO(ravicious): Copy stuff from DaemonService.reissueGatewayCerts in order to handle expired certs.
+	cert, err := cluster.ReissueAppCert(ctx, client, app)
+	return cert, trace.Wrap(err)
+}
+
+// GetDialOptions returns ALPN dial options for the profile.
+func (p *appProvider) GetDialOptions(ctx context.Context, profileName string) (*vnet.DialOptions, error) {
+	profile, err := p.clientStore.GetProfile(profileName)
```

Comment:
i wonder if you need clientStore and profile loading in here or you can reuse the `lib/teleterm/clusters` stuff effectively